### ISSUE TITLE
Check for third arg of wp_redirect() / wp_safe_redirect()

### ISSUE THREAD

**Issue by GaryJones** [Automattic/VIP-Coding-Standards#433]
## What problem would the enhancement address for VIP?
Debugging redirects can be hard as there are so many levels where a redirect may happen. 

WordPress 5.1 and later adds an `X-Redirect-By` header when `wp_redirect()` or `wp_safe_redirect()` is used, with a default value of `WordPress`, but this can be customised to any string value by adding a third argument to those function calls. This would make debugging redirects easier.

## Describe the solution you'd like

Depending on whether we could make this mandatory, this would be an Error or Warning if the argument was missing. There would be no requirement on what the string should say, but the expectation is that it would be representative of the client's theme or plugin name.

## What code should be reported as a violation?

```php
wp_safe_redirect( $location, 302 );
wp_redirect( 'https://...', 301 );
```

## What code should *not* be reported as a violation?

```php
wp_safe_redirect( $location, 302, 'Client Z Redirects plugin' );
wp_redirect( 'https://...', 301, 'Foo theme' );
```

**Comment by GaryJones** [comment]
I just wanted to note, that as well as checking for the third arg, I spent time adding some extra checks to see if the third arg was a string (if present), and the second arg was an integer. Here's the working code:

```php
public function process_parameters( $stackPtr, $group_name, $matched_content, $parameters ) {
		if ( count( $parameters ) > 2 ) {
			$actual_third_param = $this->phpcsFile->findNext( Tokens::$emptyTokens, $parameters[3]['start'], null, true );
			// Check 3rd parameter is a string.
			if ( ! in_array( $this->tokens[ $actual_third_param ]['code'], Tokens::$stringTokens, true ) ) {
				$message = 'Third parameter for "%s" should be a string.';
				$data    = [ $matched_content ];
				$this->phpcsFile->addError( $message, $stackPtr, 'ThirdParameterNotAString', $data );
			}
		} else {
			// No third parameter.
			$message = 'Missing recommended third parameter for "%s". This helps to identify the source of the redirect.';
			$data    = [ $matched_content ];
			$this->phpcsFile->addWarning( $message, $stackPtr, 'MissingThirdParameter', $data );
		}

		if ( count( $parameters ) > 1 ) {
			$actual_second_param = $this->phpcsFile->findNext( Tokens::$emptyTokens, $parameters[2]['start'], null, true );
			// Check 2nd parameter is an integer.
			if ( T_LNUMBER !== $this->tokens[ $actual_second_param ]['code'] ) {
				$message = 'Second parameter for "%s" should be an integer.';
				$data    = [ $matched_content ];
				$this->phpcsFile->addError( $message, $stackPtr, 'SecondParameterNotAnInteger', $data );

				return;
			}

			// Check 2nd parameter is 201 or 300–308.
			if ( ! in_array( (int) $this->tokens[ $actual_second_param ]['content'], $this->valid_status_codes_for_location_header, true ) ) {
				$message = 'Second parameter for "%s" should be 201 or in the range 300–308. %s found.';
				$data    = [ $matched_content, $this->tokens[ $actual_second_param ]['content'] ];
				$this->phpcsFile->addError( $message, $stackPtr, 'SecondParameterNotAValidInteger', $data );
			}

		}
	}
```

All good it would seem - with unit tests like the following:

```php
wp_redirect( 'location', 302, 'Foo1' ); // OK.
wp_redirect(
	'location',     302,'Foo2'
); // OK.
wp_redirect( 'location', 302 ); // Warning - third arg missing.
wp_redirect( 'location', 302, 301 ); // Error - third arg not a string.
wp_redirect('location','306','Foo3'); // Error - second arg not a string.

wp_safe_redirect( 'location', 302, 'Foo4' ); // OK.
wp_safe_redirect(
	'location',     302,'Foo'
); // OK.
wp_safe_redirect( 'location', 306 ); // Warning - third arg missing.
wp_safe_redirect( 'location', 302, 301 ); // Error - third arg not a string.
wp_safe_redirect('location',"307",'Foo5'); // Error - second arg not an integer.
wp_safe_redirect(   'location','a308',     true); // Error x 2 - second arg not an integer, third arg is not a string.
wp_safe_redirect( 'location', 410, 'Foo6' ); // Error - second arg not a valid integer.
```

But then I came up with unit tests like:

```php
wp_safe_redirect( 'location', (int)  "308", 'Foo 8' ); // OK.
wp_safe_redirect( 'location', 404 - 100, 'Foo 9' ); // OK.
$my_status = 123;
$my_header = 'Foo7';
wp_safe_redirect( 'location', $my_status, $my_header . ' ' . $my_status );
```

And I realised that must be a load of ways to pass in an integer value, other than just via a `T_LNUMBER`, all of which would get a false positive - static analysis would be better here.

As such, I'll scale this all the way back to just checking if the third arg exists or not.

**Comment by rebeccahum** [comment]
I like the idea and it would definitely be a better practice... but at the same time, I think this would add a lot of noise since typically the third parameter isn't filled out as it's optional.

**Comment by GaryJones** [comment]
Right, but for the benefit of Future Us and Future Them in terms of debugging, I think it would be worth encouraging / requiring this third arg.

I don't think it would add too much noise - these function calls aren't going to appear that often (compared with, say, the general use of superglobals) - they are quite specific.

**Comment by GaryJones** [comment]
@mdbitz,

> Can I ask what benefit this gives us as a team?

Having the third arg would help with debugging when redirects are happening when they perhaps shouldn't be. Not having it would mean it indicates that `WordPress` is doing the redirect, which can lead developers down the wrong road.

> I am in the boat where I think this is adding additional noise with not a large return on investment.

As already stated - this wouldn't be much noise: calls to `wp_redirect()` and `wp_safe_redirect()` are a lot fewer in number than accessing globals. The investment into the sniff is done in #440, and the investment from client developers is to be encouraged to follow a good practice. 

Is it essential to getting their code working to meet their business solution? No, but that's why it would be a warning, not an error. I don't see it as noise, but as something that has a practical benefit.

> As you outlined it can improve debugging but I don't see this as something we should enforce on our clients.

As a Warning, they could ignore it, but they can also choose to ignore pretty much all Warnings and Errors anyway if they really wish to; there's very little we can actually enforce, so this would be no different.

> To elaborate I would not want to see this warning show in the PHP CI bot feedback when reviewing code.

I disagree. Every instance of the two function calls that don't have the third arg, I comment on to suggest adding it and why.

----

@emrikol,

> This is cool! I had no idea this was an option.

Right - it's useful, has a benefit, and I think we should be encouraging clients to use it, and one way of doing that is via a Warning.

> I love the idea of it being a warning, but low enough that it doesn't show up in the CI bot.

The bot uses severity=1, so it would show up, and that's my intent here. I'd be happy to reduce the severity in #440 to be lower.

> Related, would auto-filtering x_redirect_by for any sandboxed requests (with a backtrace or something?) be beneficial? If this is only for internal debugging, it might be better to add to our tooling--but providing users the knowledge that they can do this might also benefit them and their debugging.

That would be something to investigate, but equally, the third arg is a core feature, added for the purpose of everyone being able to identify the source of redirects (including in production), so any pre-filtering of the value (for sandboxes) doesn't negate the suggestion of encouraging everyone to use the third arg directly.

**Comment by GaryJones** [comment]
@rebeccahum With the Code Quality program, and the option to drop the severity down, do you still feel that this would be noise, or a way to automate some common guidance?

**Comment by rebeccahum** [comment]
@GaryJones This still feels like noise to me — I think it would be "nice to have" and it would be a good practice during debugging but at the same time, the impact versus noise seems to be minimal IMHO.

**Comment by GaryJones** [comment]
OK, closing 👍
